### PR TITLE
Add LD_PRELOAD for asan RT if used

### PR DESCRIPTION
**Pull request by cgmillette**
Clang / Hipcc  C++ toolchains building Tensile may have added ASAN options to support the feature. Binaries must either link ASAN libraries or pre-load them.

Tensile doesn't officially support the ASAN feature yet, so this patch implements the pre-load route.

**Consumers of Tensile who wish to use ASAN shall set the environment variable ENABLE_ADDRESS_SANITIZER=true for Tensile to build without error using the new toolchain options.**

### CODE REVIEW

File: Tensile/cmake/TensileConfig.cmake [hunk 1]
```diff
@@ -224,6 +224,17 @@ function(TensileCreateLibraryFiles
       set(Tensile_MANIFEST_FILE_PATH "${Tensile_OUTPUT_PATH}/library/TensileManifest.txt")
       message(STATUS "Tensile_MANIFEST_FILE_PATH: ${Tensile_MANIFEST_FILE_PATH}")
 
+      if(DEFINED ENV{ENABLE_ADDRESS_SANITIZER} AND $ENV{ENABLE_ADDRESS_SANITIZER})
```
Comment (zaliu) [inline Tensile/cmake/TensileConfig.cmake:227]: A missing $ for ENV?

**Review by TorreZuk — COMMENTED** [review]
Fine by me if it all passes.  Maybe move the reading inside the if and fetch to new variable, i.e.
   if(defined env{ENABLE...))
       set(use_sanitizer "$ENV{ENABLE... }"
       Is it set as  bool or 1,0 in the script?   If so then add another if( use_sanitizer)
       
So adding the flags for compile and link would be a separate PR, but I may be able to remove my install.sh hack from develop.   I still don't trust the cmake CMAKE_CXX_FLAGS from carrying into our cmake project but if it all builds and links.   I found the env variables CXX_FLAGS were required to get linkage without errors in tensile.   Loading a .so before all executables seems not much better but at least it is inside tensile.   We definitely need the asan support in mainline compiler so we can verify better.

File: Tensile/cmake/TensileConfig.cmake [hunk 2]
```diff
@@ -224,6 +224,19 @@ function(TensileCreateLibraryFiles
       set(Tensile_MANIFEST_FILE_PATH "${Tensile_OUTPUT_PATH}/library/TensileManifest.txt")
       message(STATUS "Tensile_MANIFEST_FILE_PATH: ${Tensile_MANIFEST_FILE_PATH}")
 
+      if(DEFINED ENV{ENABLE_ADDRESS_SANITIZER})
+        if($ENV{ENABLE_ADDRESS_SANITIZER})
```
[the remark this answers is not in the document]
Comment (cgmillette) [reply Tensile/cmake/TensileConfig.cmake:228]: I will check to see if cmake causes an issue for accessing undefined environment variables.

The ENABLE_ADDRESS_SANITIZER variable is something that apparently is part of the compute-rocm-dkms-address-sanitizer build environment (example docker: compute-artifactory.amd.com:5000/rocm-plus-docker/compute-rocm-dkms-address-sanitizer:149-sles-rocblas-failed). This environment has build scripts for each library, which invoke the library build interface with appropriate sanitizer flags where supported.

For example, rocblas build interface (install.sh) is invoked with --address-sanitizer flag if the compute-rocm-dkms-address-sanitizer build environment contains ENABLE_ADDRESS_SANITIZER=true 

For now, Tensile doesn't have official support for an address sanitizer flag in its build interface, so we can get by checking the env variable ENABLE_ADDRESS_SANITIZER for now. At least the mathlibs builds will be able to proceed. Outside of this, consumers of Tensile using ASAN will need to be aware of the flag. Once the Tensile build interface features its own address sanitizer flag we can manage with that instead, which will clean things up.

[the remark this answers is not in the document]
Comment (cgmillette) [reply Tensile/cmake/TensileConfig.cmake]: <edit>: Seems cmake will just provide an empty string on undefined env. vars. Updated to simplify the check

**Review by TorreZuk — APPROVED** [review]
As I said fine by me, but as this is a runtime workaround rather than build time it is a little strange.   If this will also go to rel5.0 branch then we should remove my hack of export CXX_FLAGS in the commit to switch the tensile tag as well.   Not sure of that plan.

**Comment by zaliu** [comment]
@TorreZuk I'm not aware asan support is targeting 5.0.

**Comment by TorreZuk** [comment]
@b-sumner was hoping to show asan works ASAP but seems to me like too many things need tweaking.    We will have a new staging and master branch in a couple weeks with all this so maybe that can suffice.   Once this merges and tag is updated in develop (or before) I'll remove my hotfix changes from rocblas but as commented this runtime variable seems strange as a fix so would have hoped 5.1 could get more standard compile flag based fix.

**Review by sdquiring — APPROVED** [review]
Torre, this problem is about the runtime environment of Python.  The alternative (perhaps longer term) fix would be to add a GlobalParameter that passes the ASAN flags when compiling source kernels, and then remove the environment variables so that Python isn't run with ASAN.

**Comment by TorreZuk** [comment]
@sdquiring nothing should be triggering ASAN runtime during the build stage, so if you can figure out what flag is triggering this python usage maybe we can get them to fix their script environment as a better fix?

**Comment by cgmillette** [comment]
Inside rocBLAS cmake, when the --address-sanitizer flag is passed in, it populates CMAKE_CXX_FLAGS with the additional address sanitizer flags. Following this is when the virtual env is invoked for Tensile and dependencies. This includes the building of python msgpack module, which has a python-c or C++ component. It inherits the CMAKE_CXX_FLAGS and thus the dependency on the runtime.

**Comment by TorreZuk** [comment]
@cgmillette so maybe removing those flags before building tools than are run in tensile would have been the better solution?   I can also move the global setting of CMAKE_CXX_FLAGS to internal variables as right now triggers a ROCMChecks Warning/Error but you would have to add back these flags for the correct targets when doing the sanitizer build mode.

**Comment by cgmillette** [comment]
@TorreZuk I think that would depend on the original asan requirement - do we want all rocblas and Tensile dependencies instrumented with asan? We have put the blanket on the fire as-is, but for further changes I think we may be spinning our wheels without the formal asan requirement regarding Tensile (at least I haven't seen one yet)!

**Comment by sdquiring** [comment]
My impression is that the `CFLAGS`, `CXXFLAGS`, and `LDFLAGS` environment variables set in rocBLAS' `'install.sh` is causing the `msgpack-c` built by the Python msgpack library to build with ASAN when building the virtualenv, but not to be linked correctly.  This environment variable works around that.

**Pull request merged**